app/channel_category.go

```diff
@@ -144,14 +144,135 @@ func (a *App) UpdateSidebarCategoryOrder(userId, teamId string, categoryOrder []
 }
 
 func (a *App) UpdateSidebarCategories(userId, teamId string, categories []*model.SidebarCategoryWithChannels) ([]*model.SidebarCategoryWithChannels, *model.AppError) {
-	result, err := a.Srv().Store.Channel().UpdateSidebarCategories(userId, teamId, categories)
+	updatedCategories, originalCategories, err := a.Srv().Store.Channel().UpdateSidebarCategories(userId, teamId, categories)
 	if err != nil {
 		return nil, model.NewAppError("UpdateSidebarCategories", "app.channel.sidebar_categories.app_error", nil, err.Error(), http.StatusInternalServerError)
 	}
 
 	message := model.NewWebSocketEvent(model.WEBSOCKET_EVENT_SIDEBAR_CATEGORY_UPDATED, teamId, "", userId, nil)
 	a.Publish(message)
-	return result, nil
+
+	a.muteChannelsForUpdatedCategories(userId, updatedCategories, originalCategories)
+
+	return updatedCategories, nil
+}
+
+func (a *App) muteChannelsForUpdatedCategories(userId string, updatedCategories []*model.SidebarCategoryWithChannels, originalCategories []*model.SidebarCategoryWithChannels) {
+	var channelsToMute []string
+	var channelsToUnmute []string
+
+	// Mute or unmute all channels in categories that were muted or unmuted
+	for i, updatedCategory := range updatedCategories {
+		originalCategory := originalCategories[i]
+
+		if updatedCategory.Muted && !originalCategory.Muted {
+			channelsToMute = append(channelsToMute, updatedCategory.Channels...)
+		} else if !updatedCategory.Muted && originalCategory.Muted {
+			channelsToUnmute = append(channelsToUnmute, updatedCategory.Channels...)
+		}
+	}
+
+	// Mute any channels moved from an unmuted category into a muted one and vice versa
+	channelsDiff := diffChannelsBetweenCategories(updatedCategories, originalCategories)
+	if len(channelsDiff) != 0 {
+		makeCategoryMap := func(categories []*model.SidebarCategoryWithChannels) map[string]*model.SidebarCategoryWithChannels {
+			result := make(map[string]*model.SidebarCategoryWithChannels)
+			for _, category := range categories {
+				result[category.Id] = category
+			}
+
+			return result
+		}
+
+		updatedCategoriesById := makeCategoryMap(updatedCategories)
+		originalCategoriesById := makeCategoryMap(originalCategories)
+
+		for channelId, diff := range channelsDiff {
+			fromCategory := originalCategoriesById[diff.fromCategoryId]
+			toCategory := updatedCategoriesById[diff.toCategoryId]
+
+			if toCategory.Muted && !fromCategory.Muted {
+				channelsToMute = append(channelsToMute, channelId)
+			} else if !toCategory.Muted && fromCategory.Muted {
+				channelsToUnmute = append(channelsToUnmute, channelId)
+			}
+		}
+	}
+
+	for _, channelId := range channelsToMute {
+		_, err := a.setChannelMuted(channelId, userId, true)
```

Comment:
I realize there are no websocket bulk updates, but I think we can still do the DB updates in bulk. And split out the websocket updates to a separate step.

- Get all channel members from a set of channels - similar to what we have `ChannelStore.(GetMembersByIds)`, we can have `ChannelStore.(GetMembersByChannels)`.
- Toggle the mute state in a loop
- And we already have `ChannelStore.(UpdateMultipleMembers)`!
- And then finally broadcast the ws message in a loop and invalidate cache.

That reduces the number of DB queries to 2 for every set of muted and unmuted channels. So 4 in total, instead of n.

Reply:
Yeah, I can take a look at doing that. It probably negates some of the other cleanup I did elsewhere, but it'll at least be ready if we add bulk websocket events in the future.

Reply:
Updated. I didn't realize we had an UpdateMultipleMembers store method since that tends to involve one database hit per item anyway.

Reply:
> if we add bulk websocket events in the future.

Yep, this is something I plan to investigate. But this also needs support from the client side to really "understand" a bulk event.